### PR TITLE
LinuxPlumbers2021: Add info on bpf_xdp_output and bpf_perf_event_output helpers

### PR DESCRIPTION
**Pull request by chaudron**

### CODE REVIEW

**Comment by chaudron** [comment]
Added simple slide on copy helpers not sure if we need more, as there is not more to tel :) Not sure how the slide looks as my tool chains does not seem to work :(

File: conference/LinuxPlumbers2021/xdp-multi-buff.org [hunk 1]
```diff
@@ -198,6 +198,10 @@ BPF_CALL_2(bpf_xdp_adjust_data, struct xdp_buff *, xdp, u32, offset)
 }
 #+end_src
 
+** Slide: XDP multi-buffers: new eBPF helpers (4/4)                  :export:
+- /bpf_xdp_output/ and /bpf_perf_event_output/:
+  - Helpers are updated and can copy the content of all buffers
+
```
Comment (LorenzoBianconi) [inline conference/LinuxPlumbers2021/xdp-multi-buff.org:253]: @chaudron maybe you can add some code snippet here, what do you think?

Comment (chaudron) [reply conference/LinuxPlumbers2021/xdp-multi-buff.org:253]: Code snippet on how to use the helper? That I can do, but adding the changes to the code makes no sense as it's not well contained.

Comment (LorenzoBianconi) [reply conference/LinuxPlumbers2021/xdp-multi-buff.org:253]: @chaudron what I mean is try to summarize what you changed in the upstream code, I think you can take a look to the presentation to see what I mean. Btw to add code snippet you just need the #+begin_src C and #+end_src C tags

**Comment by chaudron** [comment]
Added code to the slide, but very minimal to make it fit, and show the old and new behaviour

File: conference/LinuxPlumbers2021/xdp-multi-buff.org [hunk 2]
```diff
@@ -226,6 +226,31 @@ out:
...
+                return 0;
+        }
+        ...
+        /* If we need to copy data from the base buffer do it */
+        if (off < base_len) {
+                ...
+        }
+        /* Copy any remaining data from the fragments */
+        for (i = 0; len && i < sinfo->nr_frags; i++) {
+                ...
```
Comment (LorenzoBianconi) [inline conference/LinuxPlumbers2021/xdp-multi-buff.org:248]: @chaudron maybe you can add some code here

File: conference/LinuxPlumbers2021/xdp-multi-buff.org [hunk 3]
```diff
@@ -226,6 +226,31 @@ out:
...
+                                  unsigned long off, unsigned long len)
+{
+        if (likely(!xdp_buff_is_mb(xdp))) {
+                memcpy(dst_buff, xdp->data + off, len);
+                return 0;
+        }
+        ...
+        /* If we need to copy data from the base buffer do it */
+        if (off < base_len) {
+                ...
```
Comment (LorenzoBianconi) [inline conference/LinuxPlumbers2021/xdp-multi-buff.org:244]: @chaudron maybe you can add some comment here

File: conference/LinuxPlumbers2021/xdp-multi-buff.org [hunk 4]
```diff
@@ -226,6 +226,31 @@ out:
...
+        }
+        /* Copy any remaining data from the fragments */
+        for (i = 0; len && i < sinfo->nr_frags; i++) {
+                ...
+        }
+      return 0;
```
Comment (LorenzoBianconi) [inline conference/LinuxPlumbers2021/xdp-multi-buff.org]: can you please fix the code indentation?

Comment (chaudron) [reply conference/LinuxPlumbers2021/xdp-multi-buff.org]: Done!

**Pull request merged**